libs/sdk-ui-dashboard/src/model/state/_infra/objRefMap.ts

```diff
@@ -0,0 +1,201 @@
+// (C) 2021 GoodData Corporation
+import {
+    IAttributeDisplayFormMetadataObject,
+    IAttributeMetadataObject,
+    ICatalogAttribute,
+    ICatalogDateAttribute,
+    IMetadataObject,
+} from "@gooddata/sdk-backend-spi";
+// (C) 2021 GoodData Corporation
+import { Identifier, isIdentifierRef, ObjectType, ObjRef } from "@gooddata/sdk-model";
+
+/**
+ * Configuration for the ObjRefMap.
+ *
+ * @internal
+ */
+export type ObjRefMapConfig<T> = {
+    /**
+     * Function that extracts `ref` from object
+     */
+    readonly refExtract: (obj: T) => ObjRef;
+
+    /**
+     * Function that extracts `id` from object
+     */
+    readonly idExtract: (obj: T) => Identifier;
+
+    /**
+     * Function that extracts `uri` from object
+     */
+    readonly uriExtract: (obj: T) => string;
+
+    /**
+     * Indicates whether strict idRef type-checking is desired. Some backends (e.g. tiger) have identifier
+     * refs use combination of `id` and `type` and have type-level `id` uniqueness constraints. On those backends,
+     * strict type checks are essential to correctly match objects.
+     *
+     * On other backends, the `type` information coming in the idRef is superfluous and should not be influencing
+     * anything.
+     */
+    readonly strictTypeCheck: boolean;
+
+    /**
+     * Type of object stored in the map.
+     */
+    readonly type: ObjectType;
+};
+
+/**
+ * Utility class that assists with type-agnostic mapping of metadata objects by ObjRef.
+ *
+ * Problem
+ * =======
+ *
+ * The challenges with ObjRef's start in context of backend that supports both uri and id ref (e.g. bear) and the client.
+ *
+ * Backend according to contract creates one type of ref - uri ref - so that is fine. However when instances of `ref` are created
+ * by the client code and are passed in through the public API (as is the case with the dashboard component) - problems start.
+ *
+ * For clients it is often more convenient to use ID refs.. because they are transferable across workspaces and because
+ * they appear in the catalog export.
+ *
+ * Doing strict ref-to-ref matching between user input and the data stored in state will result in no matches because
+ * the types of ref's do not match.
+ *
+ * ---
+ *
+ * This class addresses the problem by having the `get` method check the type of ObjRef first and then perform
+ * lookups into either id to item or uri to item mapping.
+ *
+ * @internal
+ */
+export class ObjRefMap<T> {
+    public readonly [Symbol.toStringTag]: string = "ObjRefMap";
+    public size: number = 0;
```

Comment:
We should probably mark this readonly

Reply:
cannot.. the size is updated during fromItems(); i have removed the clear function though as these maps are supposed to be readonly after initial populate

Reply:
🤦  my bad, it was supposed to be a "public getter-only" property, but IE, so we cannot use it.